File: migrate_titles.cpp
#include "nx/cli/application.hpp"
#include "nx/core/note.hpp"
#include <iostream>
#include <vector>
#include <set>

using namespace nx;
using namespace nx::core;

// Function to check if content already starts with the current title
bool contentStartsWithTitle(const Note& note, const std::string& stored_title) {
    if (stored_title.empty() || note.content().empty()) {
        return false;
    }
    
    std::string first_line;
    size_t end_of_line = note.content().find('\n');
    if (end_of_line != std::string::npos) {
        first_line = note.content().substr(0, end_of_line);
    } else {
        first_line = note.content();
    }
    
    // Remove leading/trailing whitespace
    auto start = first_line.find_first_not_of(" \t\r");
    if (start == std::string::npos) return false;
    auto end = first_line.find_last_not_of(" \t\r");
    first_line = first_line.substr(start, end - start + 1);
    
    // Check if it matches the title (with or without markdown heading)
    return first_line == stored_title || 
           first_line == "# " + stored_title ||
           first_line == "## " + stored_title;
}

int main(int argc, char* argv[]) {
    try {
        // Initialize the application
        nx::cli::Application app;
        auto init_result = app.initialize();
        if (!init_result.has_value()) {
            std::cerr << "Failed to initialize application: " << init_result.error().message() << std::endl;
            return 1;
        }
        
        std::cout << "=== Note Title Migration Tool ===" << std::endl;
        std::cout << "This will update notes to ensure their title matches the first line of content." << std::endl;
        
        // Get all notes
        auto note_ids_result = app.noteStore().getAllNoteIds();
        if (!note_ids_result.has_value()) {
            std::cerr << "Failed to get note IDs: " << note_ids_result.error().message() << std::endl;
            return 1;
        }
        
        auto note_ids = *note_ids_result;
        std::cout << "Found " << note_ids.size() << " notes to analyze..." << std::endl;
        
        std::vector<NoteId> notes_to_migrate;
        std::set<std::string> problematic_titles;
        
        // Analyze each note
        for (const auto& note_id : note_ids) {
            auto note_result = app.noteStore().load(note_id);
            if (!note_result.has_value()) {
                std::cerr << "Warning: Failed to load note " << note_id.toString() << std::endl;
                continue;
            }
            
            auto note = *note_result;
            std::string stored_title = note.metadata().title();
            std::string derived_title = note.title();  // This uses our new logic
            
            // Skip notebook marker notes (they start with .)
            if (stored_title.starts_with(".notebook")) {
                continue;
            }
            
            // Check if migration is needed
            bool needs_migration = false;
            std::string reason;
            
            if (stored_title.empty()) {
                needs_migration = true;
                reason = "empty stored title";
            } else if (stored_title.find("{{") != std::string::npos) {
                needs_migration = true;
                reason = "template variables in title";
            } else if (!contentStartsWithTitle(note, stored_title)) {
                needs_migration = true;
                reason = "first line doesn't match stored title";
            }
            
            if (needs_migration) {
                notes_to_migrate.push_back(note_id);
                std::cout << "  NEEDS MIGRATION: " << note_id.toString() << " - " << reason << std::endl;
                std::cout << "    Stored title: '" << stored_title << "'" << std::endl;
                std::cout << "    Derived title: '" << derived_title << "'" << std::endl;
                std::cout << "    Content preview: " << note.content().substr(0, 100) << "..." << std::endl;
                std::cout << std::endl;
            }
        }
        
        std::cout << "Analysis complete. " << notes_to_migrate.size() << " notes need migration." << std::endl;
        
        if (notes_to_migrate.empty()) {
            std::cout << "No migration needed!" << std::endl;
            return 0;
        }
        
        // Ask for confirmation
        std::cout << "Proceed with migration? (y/N): ";
        std::string response;
        std::getline(std::cin, response);
        
        if (response != "y" && response != "Y") {
            std::cout << "Migration cancelled." << std::endl;
            return 0;
        }
        
        // Perform migration
        std::cout << "Starting migration..." << std::endl;
        int migrated_count = 0;
        
        for (const auto& note_id : notes_to_migrate) {
            auto note_result = app.noteStore().load(note_id);
            if (!note_result.has_value()) {
                std::cerr << "Error loading note " << note_id.toString() << " during migration" << std::endl;
                continue;
            }
            
            auto note = *note_result;
            std::string stored_title = note.metadata().title();
            
            // Skip template notes - they need manual fixing
            if (stored_title.find("{{") != std::string::npos) {
                std::cout << "  SKIPPING template note: " << note_id.toString() << " (needs manual attention)" << std::endl;
                continue;
            }
            
            // Determine the new content
            std::string new_content;
            
            if (stored_title.empty()) {
                // For empty titles, keep content as-is (it will derive title from first line)
                new_content = note.content();
            } else {
                // Prepend the stored title as a markdown heading
                new_content = "# " + stored_title;
                if (!note.content().empty()) {
                    // If content doesn't start with newline, add one
                    if (note.content()[0] != '\n') {
                        new_content += "\n\n";
                    } else {
                        new_content += "\n";
                    }
                    new_content += note.content();
                }
            }
            
            // Update the note
            note.setContent(new_content);
            
            // Save the updated note
            auto save_result = app.noteStore().store(note);
            if (!save_result.has_value()) {
                std::cerr << "Error saving note " << note_id.toString() << ": " << save_result.error().message() << std::endl;
                continue;
            }
            
            // Update search index
            auto index_result = app.searchIndex().updateNote(note);
            if (!index_result.has_value()) {
                std::cerr << "Warning: Failed to update search index for " << note_id.toString() << std::endl;
            }
            
            migrated_count++;
            std::cout << "  MIGRATED: " << note_id.toString() << " - '" << stored_title << "'" << std::endl;
        }
        
        std::cout << "Migration complete! Successfully migrated " << migrated_count << " notes." << std::endl;
        
        // Show notes that still need manual attention
        if (migrated_count < notes_to_migrate.size()) {
            std::cout << "\nNotes requiring manual attention:" << std::endl;
            for (const auto& note_id : notes_to_migrate) {
                auto note_result = app.noteStore().load(note_id);
                if (note_result.has_value()) {
                    auto note = *note_result;
                    if (note.metadata().title().find("{{") != std::string::npos) {
                        std::cout << "  " << note_id.toString() << " - contains template variables" << std::endl;
                    }
                }
            }
        }
        
        return 0;
        
    } catch (const std::exception& e) {
        std::cerr << "Migration failed with error: " << e.what() << std::endl;
        return 1;
    }
}

File: migrate_titles.py
#!/usr/bin/env python3

import os
import re
import shutil
from datetime import datetime
from pathlib import Path

def parse_note_file(filepath):
    """Parse a note file into YAML frontmatter and content."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Split by YAML frontmatter delimiters
    parts = content.split('---\n', 2)
    if len(parts) != 3:
        return None, None, None
    
    yaml_content = parts[1].strip()
    note_content = parts[2].strip()
    
    return yaml_content, note_content, content

def extract_title_from_yaml(yaml_content):
    """Extract title from YAML frontmatter."""
    for line in yaml_content.split('\n'):
        if line.startswith('title:'):
            title = line[6:].strip()
            # Remove quotes if present
            title = re.sub(r'^["\']|["\']$', '', title)
            return title
    return ""

def get_first_line_title(content):
    """Get the first non-empty line, cleaned up."""
    if not content:
        return ""
    
    first_line = content.split('\n')[0].strip()
    # Remove markdown heading markers
    first_line = re.sub(r'^#+\s*', '', first_line)
    return first_line

def needs_migration(stored_title, first_line):
    """Check if a note needs migration."""
    if not stored_title:
        return False  # Empty titles are ok
    
    if '{{' in stored_title:
        return True  # Template variables need manual attention
    
    # Check if first line matches title (with or without markdown)
    return first_line != stored_title

def migrate_note(filepath, stored_title, yaml_content, note_content):
    """Migrate a single note file."""
    # Add title as markdown heading at the beginning
    new_content = f"# {stored_title}\n\n{note_content}"
    
    # Reconstruct full file
    full_content = f"---\n{yaml_content}\n---\n\n{new_content}\n"
    
    # Write back to file
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(full_content)

def main():
    print("=== Note Title Migration Tool ===")
    print("This will update notes to ensure their title matches the first line of content.")
    
    notes_dir = Path.home() / ".local/share/nx/notes"
    if not notes_dir.exists():
        print(f"Error: Notes directory not found at {notes_dir}")
        return 1
    
    # Create backup
    backup_dir = notes_dir.parent / f"notes_backup_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
    print(f"Creating backup at: {backup_dir}")
    shutil.copytree(notes_dir, backup_dir)
    
    # Find all note files
    note_files = list(notes_dir.glob("*.md"))
    print(f"Found {len(note_files)} notes to analyze...")
    
    if not note_files:
        print("No notes found to migrate.")
        return 0
    
    notes_to_migrate = []
    
    # Analyze each note
    for note_file in note_files:
        print(f"\nAnalyzing: {note_file.name}")
        
        yaml_content, note_content, full_content = parse_note_file(note_file)
        if yaml_content is None:
            print("  SKIP: Invalid format")
            continue
        
        stored_title = extract_title_from_yaml(yaml_content)
        first_line = get_first_line_title(note_content)
        
        print(f"  Stored title: '{stored_title}'")
        print(f"  First line: '{first_line}'")
        
        # Skip notebook markers
        if stored_title.startswith('.notebook'):
            print("  SKIP: Notebook marker file")
            continue
        
        # Check for template variables
        if '{{' in stored_title:
            print("  SKIP: Contains template variables (needs manual attention)")
            continue
        
        if needs_migration(stored_title, first_line):
            notes_to_migrate.append((note_file, stored_title, yaml_content, note_content))
            print("  NEEDS MIGRATION")
        else:
            print("  OK: Title already matches")
    
    print(f"\nAnalysis complete. {len(notes_to_migrate)} notes need migration.")
    
    if not notes_to_migrate:
        print("No migration needed!")
        return 0
    
    # Ask for confirmation
    response = input("Proceed with migration? (y/N): ")
    if response.lower() != 'y':
        print("Migration cancelled.")
        return 0
    
    # Perform migration
    print("Starting migration...")
    migrated_count = 0
    
    for note_file, stored_title, yaml_content, note_content in notes_to_migrate:
        try:
            migrate_note(note_file, stored_title, yaml_content, note_content)
            print(f"  MIGRATED: {note_file.name} - '{stored_title}'")
            migrated_count += 1
        except Exception as e:
            print(f"  ERROR: Failed to migrate {note_file.name}: {e}")
    
    print(f"\nMigration complete! Successfully migrated {migrated_count} notes.")
    print(f"Backup created at: {backup_dir}")
    
    if migrated_count > 0:
        print("\nYou can now use your notes with the new title system.")
        print(f"If something goes wrong, restore from backup: cp -r '{backup_dir}'/* '{notes_dir}'/")
    
    return 0

if __name__ == "__main__":
    exit(main())